internal/middleware/perfinsights/perfinsights.go

```diff
@@ -0,0 +1,189 @@
+	labels := make([]string, len(allLabels)+1)
+	labels[0] = methodName
+
+	for index, label := range allLabels {
+		shapeValue, ok := shape[APIShapeLabel(label)]
+		if !ok {
+			continue
+		}
+
+		switch t := shapeValue.(type) {
+		case string:
+			labels[index+1] = t
+
+		case int:
+			labels[index+1] = strconv.Itoa(t)
+
+		case uint32:
+			intValue, err := safecast.ToInt(t)
+			if err != nil {
+				log.Warn().Str("method", methodName).
+					Str("key", label).
+					Str("type", fmt.Sprintf("%T", t)).
+					Msg("ShapeBuilder could not convert int, skipping reporting")
+				return
+			}
+			labels[index+1] = strconv.Itoa(intValue)
+
+		case int64:
+			intValue, err := safecast.ToInt(t)
+			if err != nil {
+				log.Warn().Str("method", methodName).
+					Str("key", label).
+					Str("type", fmt.Sprintf("%T", t)).
+					Msg("ShapeBuilder could not convert int, skipping reporting")
+				return
+			}
+			labels[index+1] = strconv.Itoa(intValue)
+
+		default:
+			spiceerrors.DebugAssert(func() bool { return false }, "unsupported type %T", t)
+			log.Warn().Str("method", methodName).
+				Str("key", label).
+				Str("type", fmt.Sprintf("%T", t)).
+				Msg("ShapeBuilder returned an unsupported type, skipping reporting")
+			return
+		}
+	}
+
+	o := APIShapeLatency.WithLabelValues(labels...)
+	if exemplarObserver, ok := o.(prometheus.ExemplarObserver); ok {
+		if spanCtx := trace.SpanContextFromContext(ctx); spanCtx.HasTraceID() {
+			traceID := prometheus.Labels{"traceID": spanCtx.TraceID().String()}
+			exemplarObserver.ObserveWithExemplar(duration.Seconds(), traceID)
+			return
+		}
+	}
+
+	o.Observe(duration.Seconds())
+}
+
+type reporter struct {
+	isEnabled bool
+}
+
+func (r *reporter) ServerReporter(ctx context.Context, callMeta interceptors.CallMeta) (interceptors.Reporter, context.Context) {
+	if !r.isEnabled {
+		return &interceptors.NoopReporter{}, ctx
+	}
+
+	_, methodName := grpcutil.SplitMethodName(callMeta.FullMethod())
+	ctx = contextWithHandle(ctx)
+	return &serverReporter{ctx: ctx, methodName: methodName}, ctx
+}
+
+type serverReporter struct {
+	interceptors.NoopReporter
+	ctx        context.Context
+	methodName string
+}
+
+var ctxKey = ctxkey.NewBoxedWithDefault[ShapeBuilder](nil)
+
+func (r *serverReporter) PostCall(err error, duration time.Duration) {
+	// Do not report if the call resulted in an error.
+	if err != nil {
+		return
+	}
+
+	builder := ctxKey.Value(r.ctx)
+	spiceerrors.DebugAssertNotNil(builder, "ShapeBuilder should not be nil")
+
+	// If not testing and nil, simply skip.
+	if builder == nil {
+		log.Warn().Str("method", r.methodName).
+			Msg("ShapeBuilder is nil, skipping reporting")
+		return
+	}
+
+	shape := builder()
+	ObserveShapeLatency(r.ctx, r.methodName, shape, duration)
+}
+
+// UnaryServerInterceptor returns a gRPC server-side interceptor that provides reporting for Unary RPCs.
+func UnaryServerInterceptor(isEnabled bool) grpc.UnaryServerInterceptor {
```

Comment:
Please create a unit test that mounts this interceptor over a gRPC server and confirms a request generates the metrics